### PR TITLE
Planet descriptions

### PR DESCRIPTION
**Pull request by zonkmachine**
Fixes some of the issues described in https://github.com/pioneerspacesim/pioneer/issues/4215

* Ensure the first letter is a capital one. 4215 example:
`"rocky planet with a Argon atmosphere."`
`"Rocky planet with a argon atmosphere."`

* Ensure words that aren't the first one in a sentence doesn't begin with a capital letter.
* Ensure there's only one space between words.
`"Tiny rocky planet with a  Methane atmosphere."`
`"Tiny rocky planet with a methane atmosphere."`

* Fix a/an issue by simply dropping 'a' from 'with a'.
`"Rocky planet with a argon atmosphere."`
`"Rocky planet with argon atmosphere."`

* Make moons moons
`"Titan: Tiny ice planet with tenuous oxygen atmosphere."`
`"Titan: Tiny ice moon with tenuous oxygen atmosphere."`

* Name asteroids around planets moons
`"Amalthea: Asteroid"`
`"Amalthea: Tiny rocky moon"`

europapioneer

### CODE REVIEW

**Comment by RacerBG** [comment]
@zonkmachine Thank you very much for the PR! This will adress some issues with the Bulgarian translation as well.

As for fixing the a/an situation I would suggest the following solution (the sentence is an example):

"Tiny rocky planet with an atmosphere composed of methane."

**Comment by zonkmachine** [comment]
> "Tiny rocky planet with an atmosphere composed of methane."

Yes, something along those lines would work.

**Comment by Web-eWorks** [comment]
> `"Medium, highly volcanic, rocky planet with a very dense nitrogen atmosphere."`

Adding a comma here unfortunately creates a very stilted and artificial-looking sentence. There is most likely a much better way to rephrase the sentence and still express the information, but for now I consider it acceptable in the without-comma form.

**Comment by zonkmachine** [comment]
pioneerplanetdescription

Trying to iron out the a/an bug. It feels like I'm just pushing the error around. A whole lot of code changes but no great improvement. Maybe it would be better to just add whatever can be added as raw information in the table below. Then there would be no question as to why the language is dry.

```
Atmosphere: CO2
Pressure: 3 psi (space suit required)
```

**Comment by Zireael07** [comment]
Well, it's a tradeoff between having text that looks natural and text that can be generated easily. In my own games, I've given up on trying to achieve it in English (I hand write dialogue in English and use a fictional language for the game where I insisted on generating it; for my space game and the atmosphere, I simply list the chemical symbols, like CO2 in your example). Too many corner cases. A/An can in theory be differentiated by checking the first letter of the next word, but again, there are corner cases....

**Comment by nozmajner** [comment]
What about just ditching the a/an?
"Rocky planet with argon atmosphere"
sounds perfectly serviceable to me. Can't right now check what kind of other phrasing we have though.

**Comment by zonkmachine** [comment]
> What about just ditching the a/an?

That clears the text up brilliantly!

File: data/lang/core/en.json
Comment (zonkmachine) [inline data/lang/core/en.json:1802]: This could be changed to `"WITH"` but I see no good reason to do this at this point if the code is going to be rewritten anyway.

**Comment by Bodasey** [comment]
Please be careful with the word "normal" - we know the atmospheres of only seven planets and one moon and some atmospheric components of a very few exoplanets.

The main component of Mars' atmosphere is CO2, not oxygen. (Earlier) Life there is still a matter of research, or fiction, but not confirmed.

In cosmic dimensions, earth's atmosphere with a lot of oxygen is seen more an exception than normal, especially with methane in it at the same time.

psi? I have never heard about this unit - please use SI units.

**Comment by nozmajner** [comment]
Mars is terraformed in Pioneer's fiction, and it is the capital even.

**Comment by zonkmachine** [comment]
> Please be careful with the word "normal" - we know the atmospheres of only seven planets and one moon and some atmospheric components of a very few exoplanets.
> 

`Normal` was a test to make a/an in English work. That whole picture was a description of how bad my attempts were going. If I ever of get a bunch of pesetas I'm going to Mars to stay.

**Comment by zonkmachine** [comment]
Some examples of what it looks like now

marsinpioneer
earthinpioneer
newhopeinpioneer

**Comment by Bodasey** [comment]
Earth is - by definition - a "terrestric world". Oceanic worlds are (almost) completely covered by water. (for scientists)

**Comment by zonkmachine** [comment]
I had a go at naming moons as such. I haven't used string interpolation to select between planet/moon but have duplicated the strings with rock in them and replaced with moon. I also renamed asteroids that orbit a planet moons. Are there asteroids orbiting stars?

europapioneer

**Comment by Zireael07** [comment]
@Bodasey: I think the word is "terrestrial", not "terrestric". Or just do as most exoplanet researchers do and use "Earth-like"

**Comment by zonkmachine** [comment]
I'm also about to make the code recognize dwarf planets but it would take some refactoring. I think we have a chance of getting this into the release coming up in three weeks. I'm thinking it's best to leave this PR pretty much as it is and wait till after the release with the rest. The translation work at this point is pretty straight forward and a lot of it is just juggling white space around.

**Review by Web-eWorks — COMMENTED** [review]
I would suggest simply migrating the 'a / an' part of the string into the atmosphere strings themselves rather than getting rid of 'a' entirely. The article 'a' logically binds to the noun rather than the preposition anyways (and I'm assuming in other languages this would result in nicer translations anyways), so the only thing you have to worry about here is any other existing code that uses these strings.

The best way to handle that would be to define another copy of these strings with the 'a/an' prefix, leaving these existing strings in upper-case to be used in other places where a Proper Noun is appropriate (e.g. in a target details pane, scan results dialog, etc.) E.g.:

```json
  "AR_ATMOSPHERE": {
    "description": "",
    "message": "Argon atmosphere"
  },
  "AN_AR_ATMOSPHERE": {
    "description": "",
    "message": "an argon atmosphere"
  },
```

Apologies I didn't get this feedback to you earlier before doing all of the work you've done already - I do quite like where this PR is going and am a fan of the changes so far!

**Comment by zonkmachine** [comment]
> so the only thing you have to worry about here is any other existing code that uses these strings.

The specific gas atmosphere strings doesn't seem to be used anywhere else.

**Comment by zonkmachine** [comment]
@Web-eWorks I applied all your suggestions apart from keeping the upper case on the atmosphere strings without an article. I would need to add extra string manipulation to set it to lower case or it would again come out as upper case in the middle of a sentence. In the second image below, the text would read: **'a very dense Sulfuric atmosphere'**
I have also introduced articles with the atmosphere density description.

marspioneer

venuspioneer

**Comment by zonkmachine** [comment]
> e a Proper Noun is appropriate (e.g. in a target details pane, scan results dialog, etc.)

This is for future stuff right? Can we leave this for now then? I'm going to continue to work on this to introduce dwarf planets among other things so I may tackle it then.

**Comment by Web-eWorks** [comment]
> I would need to add extra string manipulation to set it to lower case or it would again come out as upper case in the middle of a sentence.

Certainly; that suggestion was made under the presumption that the string would otherwise be unused and could be repurposed in other places.

> This is for future stuff right? Can we leave this for now then? I'm going to continue to work on this to introduce dwarf planets among other things so I may tackle it then.

Indeed - with the additional context above, it's clear that such strings would be better off added to e.g. the `ui-core` module instead.

On a related note, is there any code cleanup you'd like to perform before merge, or is this PR good to squash-merge?

I'm extremely happy with what I'm seeing, the example descriptions read very naturally to a native English speaker and should hopefully work fairly well when translated - well done! The code is a bit spaghetti-like with all of the special-case handling, but well within acceptable limits for the complexity involved.

In the future I'd urge moving the code for generating these descriptions into Lua where it can be made slightly cleaner with more expressive string utilities - last I checked these descriptions are not actually used inside of C++ now that all of the oldUI code is gone.

**Comment by zonkmachine** [comment]
> On a related note, is there any code cleanup you'd like to perform before merge, or is this PR good to squash-merge?

I just went over the changes and I'm good with a merge.

**Review by Web-eWorks — APPROVED** [review]

**Pull request merged**
